clippy_lints/src/direct_recursion.rs

```diff
@@ -0,0 +1,161 @@
+        // The point of this stack is to keep track of whichever functions we're inside of.
+        // Therefore, as soon as we're about to walk a function, we push its ID into the
+        // stack. And when we're done walking it, we pop it and carry on.
+        //
+        // Doing this will later allow us to very quickly check whether a function
+        // or method definition we're calling (or binding to a fn pointer) is
+        // happening in the body of a function that contains that call (or binding).
+        //
+        // We're basically doing our homework early so that we're able to have fun
+        // later :D
+        self.fn_id_stack.push(id);
+        walk_fn(self, fk, fd, b, id);
+        _ = self.fn_id_stack.pop();
+    }
+
+    fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
+        // Before running the lint, we look up the attributes of this Expr.
+        // If it has been marked with `clippy::allowed_recursion`, then
+        // we ignore it, as well as everything inside it.
+        // Someone has decided that the recursive calls within it are fine.
+        let attrs = self.cx.tcx.hir_attrs(expr.hir_id);
+
+        for attr in attrs {
+            if let Attribute::Unparsed(attr_item) = attr {
+                for segment in &attr_item.path.segments {
+                    if segment.name == sym::allowed_recursion {
+                        return;
+                    }
+                }
+            }
+        }
+        // A Path in this context could be a path to a function or a method.
+        // If it is a Path, we check if it's a Resolved path, which in this case
+        // is a definition that we can already fully see, and therefore we can
+        // ascertain whether or not it is a recursive call.
+        if let ExprKind::Path(QPath::Resolved(_, path)) = expr.kind
+            // Now we know that this Path is fully resolved. We now
+            // must check if it points to a function or a method's definition.
+            && let Res::Def(DefKind::Fn | DefKind::AssocFn, fn_path_id) = path.res
+        {
+            // 1) Now we know that the path we've found is of a function or method definition.
+            //
+            // 2) We will now check if it corresponds to the path of a function we're inside
+            // of.
+            //
+            // 3) Thankfully, we've kept track of the functions that surround us, in `self.fn_id_stack`.
+            //
+            // 4) If the path that we've captured from `expr` coincides with one of the functions
+            // in the stack, then we know we have a recursive loop.
+
+            for fn_name in &self.fn_id_stack {
+                if fn_path_id == (*fn_name).into() {
+                    span_lint(
+                        self.cx,
+                        DIRECT_RECURSION,
+                        expr.span,
+                        "this function contains a call to itself",
+                    );
+                    break;
```

Comment:
Should this be a `return` instead? :thinking:

Reply:
I've changed it to a `return` call for now. That simplifies some of the logic related to binding of function pointers :)